### PR TITLE
HBASE-22012 Space Quota: DisableTableViolationPolicy will cause cycles of enable/disable table

### PR DESCRIPTION
**Pull request by shardul-cr7**
Space Quota: Policy state is getting changed from disable to Observance after sometime automatically.

Steps:

1: Create a table with space quota policy as Disable
2: Put some data so that table state is in space quota violation
3: So observe that table state is in violation
4: Now wait for some time
5: Observe that after some time table state is changing to to Observance however table is still disabled

Solution:  it is better to not remove the entries from cache regionSizes if table is in violation and has DISABLE policy set on it to avoid this repeat cycle of enable/disable. If the entries are not removed from the cache for disable policy and violation tables the table will not go from disable to enable even though regions of the table are offline.

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 118 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 501 | master passed |
| +1 | compile | 75 | master passed |
| +1 | checkstyle | 103 | master passed |
| +1 | shadedjars | 366 | branch has no errors when building our shaded downstream artifacts. |
| +1 | javadoc | 53 | master passed |
| 0 | spotbugs | 317 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 315 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 453 | the patch passed |
| +1 | compile | 78 | the patch passed |
| +1 | javac | 78 | the patch passed |
| +1 | checkstyle | 108 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 368 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 1415 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | javadoc | 52 | the patch passed |
| +1 | findbugs | 340 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 13584 | hbase-server in the patch failed. |
| +1 | asflicense | 26 | The patch does not generate ASF License warnings. |
| | | 18490 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/572 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 0447dc6a2f5a 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-572/out/precommit/personality/provided.sh |
| git revision | master / 345c21dbe7 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/1/testReport/ |
| Max. process+thread count | 4892 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 38 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 323 | master passed |
| +1 | compile | 55 | master passed |
| +1 | checkstyle | 77 | master passed |
| +1 | shadedjars | 275 | branch has no errors when building our shaded downstream artifacts. |
| +1 | javadoc | 37 | master passed |
| 0 | spotbugs | 262 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 260 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 291 | the patch passed |
| +1 | compile | 58 | the patch passed |
| +1 | javac | 58 | the patch passed |
| +1 | checkstyle | 75 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 267 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 912 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | javadoc | 39 | the patch passed |
| +1 | findbugs | 272 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 10431 | hbase-server in the patch failed. |
| +1 | asflicense | 32 | The patch does not generate ASF License warnings. |
| | | 13820 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/572 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux a1bbdf86b2a2 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-572/out/precommit/personality/provided.sh |
| git revision | master / a5ef6b287c |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/2/testReport/ |
| Max. process+thread count | 4601 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 32 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 351 | master passed |
| +1 | compile | 57 | master passed |
| +1 | checkstyle | 88 | master passed |
| +1 | shadedjars | 297 | branch has no errors when building our shaded downstream artifacts. |
| +1 | javadoc | 36 | master passed |
| 0 | spotbugs | 266 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 264 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 334 | the patch passed |
| +1 | compile | 57 | the patch passed |
| +1 | javac | 57 | the patch passed |
| +1 | checkstyle | 87 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 299 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 1016 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | javadoc | 34 | the patch passed |
| +1 | findbugs | 267 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 9424 | hbase-server in the patch failed. |
| +1 | asflicense | 26 | The patch does not generate ASF License warnings. |
| | | 13077 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.regionserver.TestHRegionWithInMemoryFlush |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/572 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux b7879d3eb4cc 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-572/out/precommit/personality/provided.sh |
| git revision | master / 9bed419cb0 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/3/testReport/ |
| Max. process+thread count | 4444 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by joshelser** [comment]
Yep. It's on my list -- Shardul pinged me already :)

Thanks for asking, S.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java [hunk 1]
```diff
@@ -702,15 +703,53 @@ int pruneEntriesOlderThan(long timeToPruneBefore) {
     Iterator<Entry<RegionInfo,SizeSnapshotWithTimestamp>> iterator =
         regionSizes.entrySet().iterator();
     while (iterator.hasNext()) {
-      long currentEntryTime = iterator.next().getValue().getTime();
-      if (currentEntryTime < timeToPruneBefore) {
+      RegionInfo regionInfo = iterator.next().getKey();
+      long currentEntryTime = regionSizes.get(regionInfo).getTime();
+      boolean isInViolationAndPolicyDisable = isInViolationAndPolicyDisable(regionInfo.getTable());
+      // do not prune the entries if table is in violation and
+      // violation policy is disable.prune entries older than time.
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: done

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java [hunk 2]
```diff
@@ -702,15 +703,53 @@ int pruneEntriesOlderThan(long timeToPruneBefore) {
...
+   *
+   * @param tableName tableName to check.
+   * @return returns true if table is in violation and policy is disable else false.
+   */
+  private boolean isInViolationAndPolicyDisable(TableName tableName) {
+    boolean isInViolationAtTable = false;
+    boolean isInViolationAndPolicyDisable = false;
+    SpaceViolationPolicy policy = null;
+    try {
+      if (QuotaUtil.isQuotaEnabled(masterServices.getConfiguration())) {
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: removed

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java [hunk 3]
```diff
@@ -702,15 +703,53 @@ int pruneEntriesOlderThan(long timeToPruneBefore) {
...
+        // Get Current Snapshot for the given table
+        SpaceQuotaSnapshot spaceQuotaSnapshot =
+            QuotaUtil.getCurrentSnapshotFromQuotaTable(masterServices.getConnection(), tableName);
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: using ```QuotaObserverChore#getTableQuotaSnapshot(TableName)``` now.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java [hunk 4]
```diff
@@ -702,15 +703,53 @@ int pruneEntriesOlderThan(long timeToPruneBefore) {
...
+        if (spaceQuotaSnapshot != null) {
+          // check if table in violation
+          isInViolationAtTable = spaceQuotaSnapshot.getQuotaStatus().isInViolation();
+          Optional<SpaceViolationPolicy> policyAtNamespace =
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: done

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java [hunk 5]
```diff
@@ -702,15 +703,53 @@ int pruneEntriesOlderThan(long timeToPruneBefore) {
...
+              spaceQuotaSnapshot.getQuotaStatus().getPolicy();
+          if (policyAtNamespace.isPresent()) {
+            policy = policyAtNamespace.get();
+          }
+        }
+      }
+      isInViolationAndPolicyDisable =
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: done

File: hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestMasterQuotaManager.java [hunk 6]
```diff
@@ -51,6 +55,10 @@ public void testOldEntriesRemoved() {
     MasterServices masterServices = mock(MasterServices.class);
     MasterQuotaManager manager = new MasterQuotaManager(masterServices);
     manager.initializeRegionSizes();
+    HBaseTestingUtility TEST_UTIL = new HBaseTestingUtility();
+    Configuration conf = TEST_UTIL.getConfiguration();
+    conf.set(QUOTA_CONF_KEY, "false");
+    when(masterServices.getConfiguration()).thenReturn(conf);
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestMasterQuotaManager.java]: removed

File: hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java [hunk 7]
```diff
@@ -221,4 +225,30 @@ public void testTableQuotaOverridesNamespaceQuota() throws Exception {
         Bytes.toBytes("reject"));
     helper.verifyViolation(policy, tn, p);
   }
+
+  @Test
+  public void testDisablePolicyQuotaAndViolate() throws Exception {
+    TableName tableName = helper.createTable();
+    helper.setQuotaLimit(tableName, SpaceViolationPolicy.DISABLE, 2L);
+    helper.writeData(tableName, SpaceQuotaHelperForTests.ONE_MEGABYTE * 3L);
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java]: changed to 1MB.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java [hunk 8]
```diff
@@ -221,4 +225,30 @@ public void testTableQuotaOverridesNamespaceQuota() throws Exception {
...
+
+    HMaster master = TEST_UTIL.getMiniHBaseCluster().getMaster();
+    MasterQuotaManager quotaManager = master.getMasterQuotaManager();
+
+    // Sufficient time for all the chores to run.
+    Thread.sleep(5000);
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java]: done. added a wait predicate.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java [hunk 9]
```diff
@@ -221,4 +225,30 @@ public void testTableQuotaOverridesNamespaceQuota() throws Exception {
...
+
+    long timeToPrune = System.currentTimeMillis() + 11 * 60 * 1000;
+    quotaManager.pruneEntriesOlderThan(timeToPrune);
+
+    // Check if disabled table region report present in the map after retention period expired.
+    // It should be present after retention period expired.
+    for (Map.Entry<RegionInfo, Long> entry : quotaManager.snapshotRegionSizes().entrySet()) {
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java]: done

**Comment by shardul-cr7** [comment]
@joshelser , thanks for the review..did all the changes and pushed.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 36s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 33s |  master passed  |
| :green_heart: |  compile  |   0m 56s |  master passed  |
| :green_heart: |  checkstyle  |   1m 18s |  master passed  |
| :green_heart: |  shadedjars  |   4m 32s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 38s |  master passed  |
| :blue_heart: |  spotbugs  |   4m  1s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   3m 58s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   4m 52s |  the patch passed  |
| :green_heart: |  compile  |   1m  5s |  the patch passed  |
| :green_heart: |  javac  |   1m  5s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 20s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 29s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  16m 20s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 34s |  the patch passed  |
| :broken_heart: |  findbugs  |   4m 10s |  hbase-server generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 156m 59s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 35s |  The patch does not generate ASF License warnings.  |
|  |   | 214m 49s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hbase-server |
|  |  Null pointer dereference of tableQuotaSnapshot in org.apache.hadoop.hbase.quotas.MasterQuotaManager.isInViolationAndPolicyDisable(TableName, QuotaObserverChore)  Dereferenced at MasterQuotaManager.java:in org.apache.hadoop.hbase.quotas.MasterQuotaManager.isInViolationAndPolicyDisable(TableName, QuotaObserverChore)  Dereferenced at MasterQuotaManager.java:[line 757] |
|  |  Load of known null value in org.apache.hadoop.hbase.quotas.MasterQuotaManager.isInViolationAndPolicyDisable(TableName, QuotaObserverChore)  At MasterQuotaManager.java:in org.apache.hadoop.hbase.quotas.MasterQuotaManager.isInViolationAndPolicyDisable(TableName, QuotaObserverChore)  At MasterQuotaManager.java:[line 756] |
| Failed junit tests | hadoop.hbase.tool.TestBulkLoadHFilesSplitRecovery |
|   | hadoop.hbase.tool.TestSecureBulkLoadHFilesSplitRecovery |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/572 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 560cf133dc92 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-572/out/precommit/personality/provided.sh |
| git revision | master / 537c5f0660 |
| Default Java | 1.8.0_181 |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/6/artifact/out/new-findbugs-hbase-server.html |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/6/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/6/testReport/ |
| Max. process+thread count | 4540 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/6/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java [hunk 10]
```diff
@@ -694,23 +695,65 @@ public void addRegionSize(RegionInfo hri, long size, long time) {
...
+  private boolean isInViolationAndPolicyDisable(TableName tableName,
+      QuotaObserverChore quotaObserverChore) {
+    boolean isInViolationAtTable = false;
+    boolean isInViolationAtNamespace = false;
+    SpaceViolationPolicy policy = null;
+    // Get Current Snapshot for the given table
+    SpaceQuotaSnapshot tableQuotaSnapshot = quotaObserverChore.getTableQuotaSnapshot(tableName);
+    SpaceQuotaSnapshot namespaceQuotaSnapshot =
+        quotaObserverChore.getNamespaceQuotaSnapshot(tableName.getNamespaceAsString());
+    if (tableQuotaSnapshot != null) {
```
Comment (joshelser) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: What if there is both a table quota and a namespace quota (both with disable violation policy), but only the namespace quota is in violation? There is a bug in this condition (would incorrectly return that there is no violation).

File: hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java [hunk 11]
```diff
@@ -221,4 +228,32 @@ public void testTableQuotaOverridesNamespaceQuota() throws Exception {
...
+        Map<RegionInfo, Long> regionSizes = quotaManager.snapshotRegionSizes();
+        List<RegionInfo> tableRegions =
+            MetaTableAccessor.getTableRegions(TEST_UTIL.getConnection(), tableName);
+        return regionSizes.containsKey(tableRegions.get(0));
+      }
+    });
+
+    // Check if disabled table region report present in the map after retention period expired.
+    // It should be present after retention period expired.
+    Assert.assertTrue(quotaManager.snapshotRegionSizes().keySet().stream()
```
Comment (joshelser) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java]: Can you pull the lambda out of the `assertTrue` to make this a little more readable? e.g.
```
final int regionSizes = quotaManager.snapshotRegionSizes()...;
Assert.assertTrue(regionSizes > 0);
```

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java (hunk 10 above)
Comment (shardul-cr7) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: Good catch. @joshelser . Did all the changes and pushed.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java (hunk 11 above)
Comment (shardul-cr7) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestSpaceQuotaBasicFunctioning.java]: done

File: hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java [hunk 12]
```diff
@@ -694,23 +695,65 @@ public void addRegionSize(RegionInfo hri, long size, long time) {
...
+      isInViolationAtTable = tableQuotaSnapshot.getQuotaStatus().isInViolation();
+      Optional<SpaceViolationPolicy> tablePolicy = tableQuotaSnapshot.getQuotaStatus().getPolicy();
+      if (tablePolicy.isPresent()) {
+        policy = tablePolicy.get();
+      }
+    } else if (namespaceQuotaSnapshot != null) {
+      // check namespace in violation
+      isInViolationAtNamespace = namespaceQuotaSnapshot.getQuotaStatus().isInViolation();
+      Optional<SpaceViolationPolicy> namespacePolicy =
+          tableQuotaSnapshot.getQuotaStatus().getPolicy();
```
[the remark this answers is not in the document]
Comment (shardul-cr7) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotaManager.java]: done

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   2m 32s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   6m 34s |  master passed  |
| :green_heart: |  compile  |   1m  9s |  master passed  |
| :green_heart: |  checkstyle  |   1m 39s |  master passed  |
| :green_heart: |  shadedjars  |   5m  9s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 41s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 35s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 33s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 56s |  the patch passed  |
| :green_heart: |  compile  |   1m  8s |  the patch passed  |
| :green_heart: |  javac  |   1m  8s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 30s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 41s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  18m 32s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m  8s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 277m 52s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 36s |  The patch does not generate ASF License warnings.  |
|  |   | 344m 44s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.tool.TestBulkLoadHFilesSplitRecovery |
|   | hadoop.hbase.tool.TestSecureBulkLoadHFilesSplitRecovery |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/7/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/572 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 7032311cacc4 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-572/out/precommit/personality/provided.sh |
| git revision | master / 537c5f0660 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/7/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/7/testReport/ |
| Max. process+thread count | 4845 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-572/7/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by joshelser** [comment]
Looks ok now. Let me try to get this in.

**Comment by joshelser** [comment]
Required some minor changes across branch-2.2/2 and more changes in branch-2.1.

Re-running the UTs on branch-2.1 now.

**Pull request closed without merging**